libnavigation-core/src/main/java/com/mapbox/navigation/core/routealternatives/RouteAlternativesController.kt

```diff
@@ -0,0 +1,110 @@
+package com.mapbox.navigation.core.routealternatives
+
+import com.mapbox.api.directions.v5.models.DirectionsRoute
+import com.mapbox.api.directions.v5.models.RouteOptions
+import com.mapbox.base.common.logger.model.Message
+import com.mapbox.navigation.base.route.RouteAlternativesOptions
+import com.mapbox.navigation.core.directions.session.DirectionsSession
+import com.mapbox.navigation.core.directions.session.RoutesRequestCallback
+import com.mapbox.navigation.core.routeoptions.RouteOptionsUpdater
+import com.mapbox.navigation.core.trip.session.TripSession
+import com.mapbox.navigation.navigator.internal.MapboxNativeNavigator
+import com.mapbox.navigation.utils.internal.LoggerProvider.logger
+import com.mapbox.navigation.utils.internal.MapboxTimer
+import com.mapbox.navigation.utils.internal.ThreadController
+import kotlinx.coroutines.cancelChildren
+import kotlinx.coroutines.launch
+import java.util.concurrent.CopyOnWriteArraySet
+
+internal class RouteAlternativesController(
+    private val options: RouteAlternativesOptions,
+    private val navigator: MapboxNativeNavigator,
+    private val directionsSession: DirectionsSession,
+    private val tripSession: TripSession,
+    private val routeOptionsUpdater: RouteOptionsUpdater
+) {
+    private val jobControl = ThreadController.getMainScopeAndRootJob()
+
+    private val mapboxTimer = MapboxTimer()
+    private val observers = CopyOnWriteArraySet<RouteAlternativesObserver>()
+    private var currentRequestId: Long? = null
+
+    fun register(routeAlternativesObserver: RouteAlternativesObserver) {
+        val needsToStartTimer = observers.isEmpty()
+        observers.add(routeAlternativesObserver)
+        if (needsToStartTimer) {
+            mapboxTimer.restartAfterMillis = options.intervalMillis
+            mapboxTimer.startTimer {
+                requestRouteAlternatives()
+            }
+        }
+    }
+
+    fun unregister(routeAlternativesObserver: RouteAlternativesObserver) {
+        observers.remove(routeAlternativesObserver)
+        if (observers.isEmpty()) {
+            stop()
+        }
+    }
+
+    fun unregisterAll() {
+        observers.clear()
+        stop()
+    }
+
+    private fun stop() {
+        mapboxTimer.stopJobs()
+        jobControl.job.cancelChildren()
+        currentRequestId?.let { directionsSession.cancelRouteRequest(it) }
+    }
+
+    private fun requestRouteAlternatives() {
+        if (directionsSession.routes.isEmpty()) {
```

Comment:
Check if the trip session is active? A `null` route progress is an error in the options updater anyway.

Reply:
Same comment about the `NavigationSession#state` here as well.

Reply:
🤔  good call disabling this while the trip session is stopped. We specifically chose to keep "Route Refresh" enabled when the trip is stopped so the traffic will update before guidance.

But requesting alternatives, is something we only want to refresh while moving. TripSession started is a pretty good indicator that they intend on moving. Discussing a road-aware interval for the future, but that is more complicated.

Added the trip session state, updated test, updated docs.